---
title: "methyLImp2_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{methyLImp2_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(methyLImp2)
library(SummarizedExperiment)
```

In this vignette we demonstrate the usage of the *methyLImp* method for the imputation of the missing values in the methylation dataset. The vignette consists of three parts: generation of the artifical missing values into example dataset (since we need to know the true values for the performance evaluation), application of *methyLImp2* and performance evaluation (we compare imputed values with the true ones).

##Missing values generation

First, we demonstrate how we add artifical NAs into the dataset and memorize their positions so that we can later evaluate the performance of the method.

Here we use a subset of GSE199057 Gene Expression Omnibus dataset with mucosa samples from non-colon-cancer patients.  Methylation data were measured on EPIC arrays. For the sake of vignette, we reduce the number of samples from 68 to randomly sampled 24. We also restrict the dataset to the two shortest chromosomes, 18 and 21, so the total number of probes is 24 273. We refer the reader to our [simulation studies](https://github.com/annaplaksienko/methyLImp2_simulation_studies) for full size dataset results.

```{r generate artifical NAs}
#load the dataset from the package
data(beta)
print(dim(beta))

with_missing_data <- generateMissingData(beta, lambda = 3.5)
beta_with_nas <- with_missing_data$beta_with_nas
na_positions <- with_missing_data$na_positions
```

You can provide a numeric matrix as an argument for *methyLImp*. However, as many users work with SummarizedExperiment objects in the Bioconductor workflow, we will construct such object here for the sake of demonstration. 
```{r construct SE}
data(beta_meta)
beta_SE <- SummarizedExperiment(assays = list(t(beta_with_nas)), colData = beta_meta)
```

## Using methyLImp

Now, let's run *methyLImp*! You start by providing either a numeric data matrix with missing values, with samples in rows and variables (probes) in columns, or a SummarizedExperiment object, from which the first assays slot will be imputed (here we provide a numeric matrix). You also need to provide the type of your data - 450k or EPIC. These are the only two arguments you have to provide. However, there are a few other things you can tune so that the method suits your needs best: 

* specify what groups you have in your data: *methyLImp* works best when imputation is done on each group of samples independetly. Therefore, you should specify the correspondence of samples to groups. Here we've already pre-filtered the data to only have one group, so we do not use this argument here;

* specify the type of data as "user-provided": type of data (450k or EPIC) is used to split CpGs across chromosomes. Match of CpGs to chromosomes is taken from Illumina website. However, it is a part of the package and is not dynamically updated. Therefore, if you wish to provide your own match, specify "user" in the type argument and provide a data frame with in the annotation argument; 

* choose the number of cores: *methyLImp* first splits the data by chromosomes and then does imputation in parallel. By default the number of cores will be one less than total number of your physical cores, but you can decrease it. Here we use 2 since for the purpose of the vignette our dataset is small but we encourage you to increase this number; 

* set up mini-batch computation: if your dataset is quite big sample-wise, you can opt to use only a fraction of samples for the imputation to decrease the running time, say 10, 20, 50% (depending on the original number of samples). Subsample will be chosen randomly for each calculation. The bigger the subsample - the better is the performance, hence by default it is 1. An option to improve the performance but still keep the running time low is to repeat computation for a (randomly chosen) fraction of samples several times, maybe 2, 3, 5. In the paper we explore how these two tuning parameters influence the running time and the performance.

The output matrix here is the original one with imputed values instead of NAs.

Note that running time should take around 2 to 3 minutes depending on your machine.

```{r run methyLImp}
time <- system.time(beta_SE_imputed <- methyLImp2(beta_SE, 
                                               type = "EPIC", 
                                               ncores = 2,
                                               minibatch_frac = 0.5))
print(paste0("Runtime was ", round(time[3], digits = 2), " seconds."))
```

## Performance evaluation

Now we evaluate the performance of the algorithm with root mean square error (RMSE) and mean absolute error (MAE): the lower the values are, the better.

```{r evaluate performance}
performance <- evaluatePerformance(beta, t(assays(beta_SE_imputed)[[1]]), na_positions)
print(performance)
```

## Session info

```{r sessionInfo}
sessionInfo()
```